File: harness_builder/core/external_loop.py
"""
The EXTERNAL LOOP — the layer outside the harness (the pink box):

    goal ──▶ PLANNER ──▶ numbered checklist (1..N)      <- like the whiteboard
                              │
              ┌───────────────▼────────────────┐
              │  for each unchecked item:      │
              │    fresh HARNESS run on item   │  <- harness ⊃ agent loop ⊃ context
              │    per-item GATE (judge/none)  │
              │    pass → ✓ check off + note   │
              │    fail → retry, then REPLAN   │
              └───────────────┬────────────────┘
                              │ until list done | max_cycles | budget

Why this increases agent scope: the PLAN LIVES IN A FILE, not in any context
window. Each harness run is fresh and small; the checklist + completed-step
notes are the only state that crosses runs. Kill the process, run again with
the same goal — it resumes exactly where it left off.

State on disk:  <harness>/loop_state/<goal-hash>.json
"""

from __future__ import annotations

import hashlib
import json
import re
from dataclasses import dataclass, field
from pathlib import Path

from ..providers import api

PLANNER_SYSTEM = """You are a planner for an autonomous agent loop.
Decompose the GOAL into 3-8 concrete, sequential steps a specialist agent team
will execute ONE PER RUN. Each step must be independently executable given only
the goal + notes from completed steps, and independently checkable.
Respond ONLY with JSON:
{"steps": [{"title": "<imperative, one line>",
            "details": "<what exactly to do, 1-3 sentences>",
            "done_when": "<concrete, checkable completion criterion>"}]}"""

REPLAN_SYSTEM = """You are revising the remaining plan of an autonomous agent
loop after a step kept failing. Given the GOAL, COMPLETED steps, the FAILED
step and its failure diagnosis, produce a revised list of remaining steps
(you may split, reorder, reword, or route around the failure — but the revised
plan must still achieve the goal). Respond ONLY with JSON:
{"steps": [{"title": "...", "details": "...", "done_when": "..."}]}"""

STEP_JUDGE_SYSTEM = """You are a strict per-step gate in an agent loop.
Given one STEP (with its done_when criterion) and the agent team's OUTPUT for
it, decide if the step is genuinely complete. Respond ONLY with JSON:
{"passed": true|false, "note": "<if passed: 1-2 sentence factual summary of
what was accomplished, for downstream steps>", "diagnosis": "<if failed:
specific fix instructions>"}"""


def _llm_json(model: str, system: str, prompt: str, max_tokens: int = 4000) -> dict:
    provider, m = api.resolve(model)
    resp = provider.chat(model=m, system=system,
                         messages=[{"role": "user", "content": prompt}],
                         max_tokens=max_tokens)
    text = re.sub(r"```(json)?|```", "", resp.text).strip()
    return json.loads(text)


@dataclass
class Step:
    title: str
    details: str = ""
    done_when: str = ""
    status: str = "pending"      # pending | done | failed
    attempts: int = 0
    note: str = ""               # summary passed to later steps


@dataclass
class LoopState:
    goal: str
    steps: list = field(default_factory=list)
    cycles: int = 0

    @staticmethod
    def path_for(harness_dir: Path, goal: str) -> Path:
        h = hashlib.sha1(goal.encode()).hexdigest()[:10]
        return harness_dir / "loop_state" / f"{h}.json"

    @staticmethod
    def load(path: Path) -> "LoopState | None":
        if not path.exists():
            return None
        d = json.loads(path.read_text())
        st = LoopState(goal=d["goal"], cycles=d.get("cycles", 0))
        st.steps = [Step(**s) for s in d["steps"]]
        return st

    def save(self, path: Path):
        path.parent.mkdir(parents=True, exist_ok=True)
        path.write_text(json.dumps(
            {"goal": self.goal, "cycles": self.cycles,
             "steps": [vars(s) for s in self.steps]}, indent=1))

    def render(self) -> str:
        icons = {"done": "✓", "pending": "·", "failed": "✗"}
        return "\n".join(f"  {icons[s.status]} {i+1}. {s.title}"
                         + (f"  [{s.attempts} attempts]" if s.attempts > 1 else "")
                         for i, s in enumerate(self.steps))


def run_external_loop(*, spec, harness_dir: Path, goal: str,
                      run_harness_fn, fresh: bool = False,
                      trace=None) -> LoopState:
    cfg = spec.loop
    state_path = LoopState.path_for(harness_dir, goal)

    # ── PLAN (or resume) ────────────────────────────────────────────────
    state = None if fresh else LoopState.load(state_path)
    if state is None:
        print(f"[planner] decomposing goal with {cfg.planner_model} ...")
        plan = _llm_json(cfg.planner_model, PLANNER_SYSTEM, f"GOAL:\n{goal}")
        state = LoopState(goal=goal,
                          steps=[Step(**s) for s in plan["steps"]])
        state.save(state_path)
    else:
        print(f"[loop] resuming from {state_path.name} "
              f"({sum(s.status == 'done' for s in state.steps)}"
              f"/{len(state.steps)} done)")
    print("\nPLAN:")
    print(state.render(), "\n")

    # ── THE LOOP ────────────────────────────────────────────────────────
    while state.cycles < cfg.max_cycles:
        step = next((s for s in state.steps if s.status == "pending"), None)
        if step is None:
            print("\n[loop] all steps checked off — goal complete")
            print(state.render())
            return state

        idx = state.steps.index(step) + 1
        state.cycles += 1
        step.attempts += 1
        print(f"═══ CYCLE {state.cycles}/{cfg.max_cycles} · "
              f"step {idx}: {step.title} (attempt {step.attempts}) ═══")

        completed_notes = "\n".join(
            f"- ({i+1}) {s.title}: {s.note or 'done'}"
            for i, s in enumerate(state.steps) if s.status == "done")
        task = (f"OVERALL GOAL (for context — do NOT do it all now):\n{state.goal}\n\n"
                + (f"COMPLETED STEPS (build on these):\n{completed_notes}\n\n"
                   if completed_notes else "")
                + f"YOUR CURRENT STEP — do exactly this and only this:\n"
                  f"{idx}. {step.title}\n{step.details}\n"
                  f"Definition of done: {step.done_when}")

        output = run_harness_fn(task)   # fresh harness run (inner loops inside)

        # ── per-item gate ───────────────────────────────────────────────
        if cfg.step_verify == "judge":
            try:
                verdict = _llm_json(
                    spec.eval.judge_model, STEP_JUDGE_SYSTEM,
                    f"STEP:\n{step.title}\n{step.details}\n"
                    f"done_when: {step.done_when}\n\nOUTPUT:\n{output[:20000]}",
                    max_tokens=1000)
            except Exception as e:
                verdict = {"passed": False, "diagnosis": f"judge failed: {e}"}
        else:
            verdict = {"passed": True,
                       "note": output[:300]}

        if trace:
            trace.log("loop_step", step=idx, title=step.title,
                      attempt=step.attempts, passed=verdict.get("passed"))

        if verdict.get("passed"):
            step.status = "done"
            step.note = verdict.get("note", "")[:500]
            print(f"  ✓ step {idx} checked off")
        else:
            diagnosis = verdict.get("diagnosis", "")[:800]
            print(f"  ✗ step {idx} failed gate: {diagnosis[:120]}")
            if step.attempts < cfg.max_attempts_per_step:
                step.details += f"\nPREVIOUS ATTEMPT FAILED — fix: {diagnosis}"
            elif cfg.replan_on_failure:
                print("  [replan] revising remaining plan around the failure")
                try:
                    done = [s for s in state.steps if s.status == "done"]
                    plan = _llm_json(
                        cfg.planner_model, REPLAN_SYSTEM,
                        f"GOAL:\n{state.goal}\n\nCOMPLETED:\n"
                        + "\n".join(f"- {s.title}: {s.note}" for s in done)
                        + f"\n\nFAILED STEP:\n{step.title}\n\nDIAGNOSIS:\n{diagnosis}")
                    step.status = "failed"
                    state.steps = done + [step] + [Step(**s) for s in plan["steps"]]
                except Exception as e:
                    print(f"  [replan] failed ({e}); marking step failed and continuing")
                    step.status = "failed"
            else:
                step.status = "failed"

        state.save(state_path)          # ← resumability: state survives crashes
        print("\n" + state.render() + "\n")

    print(f"[loop] stopped: max_cycles ({cfg.max_cycles}) reached")
    return state

File: harness-rs/src/external_loop.rs
//! The EXTERNAL LOOP — the layer outside the harness (whiteboard's pink box):
//! goal → planner → numbered checklist → one fresh harness run per item →
//! per-item gate → check off → replan on repeated failure → done when the
//! list is done. State lives in a file, so the loop is resumable and its
//! scope exceeds any single context window.

use anyhow::{Context, Result};
use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::json;
use std::path::{Path, PathBuf};

use crate::providers::resolve;
use crate::spec::HarnessSpec;

const PLANNER_SYSTEM: &str = "You are a planner for an autonomous agent loop.\n\
Decompose the GOAL into 3-8 concrete, sequential steps a specialist agent team \
will execute ONE PER RUN. Each step must be independently executable given only \
the goal + notes from completed steps, and independently checkable.\n\
Respond ONLY with JSON:\n\
{\"steps\": [{\"title\": \"<imperative, one line>\", \"details\": \"<what exactly \
to do, 1-3 sentences>\", \"done_when\": \"<concrete, checkable completion \
criterion>\"}]}";

const REPLAN_SYSTEM: &str = "You are revising the remaining plan of an \
autonomous agent loop after a step kept failing. Given the GOAL, COMPLETED \
steps, the FAILED step and its failure diagnosis, produce a revised list of \
remaining steps that still achieves the goal. Respond ONLY with JSON:\n\
{\"steps\": [{\"title\": \"...\", \"details\": \"...\", \"done_when\": \"...\"}]}";

const STEP_JUDGE_SYSTEM: &str = "You are a strict per-step gate in an agent \
loop. Given one STEP (with its done_when criterion) and the agent team's \
OUTPUT for it, decide if the step is genuinely complete. Respond ONLY with \
JSON: {\"passed\": true|false, \"note\": \"<if passed: 1-2 sentence factual \
summary for downstream steps>\", \"diagnosis\": \"<if failed: specific fix \
instructions>\"}";

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Step {
    pub title: String,
    #[serde(default)]
    pub details: String,
    #[serde(default)]
    pub done_when: String,
    #[serde(default = "pending")]
    pub status: String, // pending | done | failed
    #[serde(default)]
    pub attempts: u32,
    #[serde(default)]
    pub note: String,
}
fn pending() -> String { "pending".into() }

#[derive(Debug, Serialize, Deserialize)]
pub struct LoopState {
    pub goal: String,
    pub steps: Vec<Step>,
    #[serde(default)]
    pub cycles: u32,
}

impl LoopState {
    fn path_for(harness_dir: &Path, goal: &str) -> PathBuf {
        // tiny stable hash (FNV-1a) — no extra crate needed
        let mut h: u64 = 0xcbf29ce484222325;
        for b in goal.bytes() {
            h ^= b as u64;
            h = h.wrapping_mul(0x100000001b3);
        }
        harness_dir.join("loop_state").join(format!("{:010x}.json", h & 0xffffffffff))
    }

    fn save(&self, path: &Path) -> Result<()> {
        std::fs::create_dir_all(path.parent().unwrap())?;
        std::fs::write(path, serde_json::to_string_pretty(self)?)?;
        Ok(())
    }

    pub fn render(&self) -> String {
        self.steps.iter().enumerate().map(|(i, s)| {
            let icon = match s.status.as_str() {
                "done" => "✓", "failed" => "✗", _ => "·" };
            let attempts = if s.attempts > 1 {
                format!("  [{} attempts]", s.attempts) } else { String::new() };
            format!("  {} {}. {}{}", icon, i + 1, s.title, attempts)
        }).collect::<Vec<_>>().join("\n")
    }

    pub fn done_count(&self) -> usize {
        self.steps.iter().filter(|s| s.status == "done").count()
    }
}

fn llm_json(model: &str, system: &str, prompt: &str,
            max_tokens: u32) -> Result<serde_json::Value> {
    let provider = resolve(model)?;
    let resp = provider.chat(system,
        &[json!({"role": "user", "content": prompt})], &[], max_tokens)?;
    let clean = Regex::new(r"```(json)?|```").unwrap()
        .replace_all(&resp.text, "").trim().to_string();
    serde_json::from_str(&clean)
        .with_context(|| format!("model returned non-JSON: {}",
            resp.text.chars().take(300).collect::<String>()))
}

fn parse_steps(v: &serde_json::Value) -> Result<Vec<Step>> {
    let steps: Vec<Step> = serde_json::from_value(v["steps"].clone())
        .context("plan JSON missing valid 'steps' array")?;
    anyhow::ensure!(!steps.is_empty(), "planner produced an empty plan");
    Ok(steps)
}

pub fn run_external_loop<F>(spec: &HarnessSpec, harness_dir: &Path, goal: &str,
                            fresh: bool, mut run_harness: F) -> Result<LoopState>
where F: FnMut(&str) -> Result<String> {
    let cfg = &spec.loop_spec;
    let state_path = LoopState::path_for(harness_dir, goal);

    // ── plan or resume ──────────────────────────────────────────────────
    let mut state = if !fresh && state_path.exists() {
        let st: LoopState = serde_json::from_str(
            &std::fs::read_to_string(&state_path)?)?;
        println!("[loop] resuming from {} ({}/{} done)",
                 state_path.file_name().unwrap().to_string_lossy(),
                 st.done_count(), st.steps.len());
        st
    } else {
        println!("[planner] decomposing goal with {} ...", cfg.planner_model);
        let plan = llm_json(&cfg.planner_model, PLANNER_SYSTEM,
                            &format!("GOAL:\n{}", goal), 4000)?;
        let st = LoopState { goal: goal.into(), steps: parse_steps(&plan)?,
                             cycles: 0 };
        st.save(&state_path)?;
        st
    };
    println!("\nPLAN:\n{}\n", state.render());

    // ── the loop ────────────────────────────────────────────────────────
    while state.cycles < cfg.max_cycles {
        let Some(i) = state.steps.iter().position(|s| s.status == "pending")
        else {
            println!("\n[loop] all steps checked off — goal complete\n{}",
                     state.render());
            return Ok(state);
        };
        state.cycles += 1;
        state.steps[i].attempts += 1;
        let (title, details, done_when, attempts) = {
            let s = &state.steps[i];
            (s.title.clone(), s.details.clone(), s.done_when.clone(), s.attempts)
        };
        println!("═══ CYCLE {}/{} · step {}: {} (attempt {}) ═══",
                 state.cycles, cfg.max_cycles, i + 1, title, attempts);

        let completed: String = state.steps.iter().enumerate()
            .filter(|(_, s)| s.status == "done")
            .map(|(j, s)| format!("- ({}) {}: {}", j + 1, s.title,
                 if s.note.is_empty() { "done" } else { &s.note }))
            .collect::<Vec<_>>().join("\n");
        let task = format!(
            "OVERALL GOAL (for context — do NOT do it all now):\n{}\n\n{}\
             YOUR CURRENT STEP — do exactly this and only this:\n{}. {}\n{}\n\
             Definition of done: {}",
            state.goal,
            if completed.is_empty() { String::new() } else {
                format!("COMPLETED STEPS (build on these):\n{}\n\n", completed) },
            i + 1, title, details, done_when);

        let output = run_harness(&task)?;   // fresh harness run

        // ── per-item gate ───────────────────────────────────────────────
        let (passed, note, diagnosis) = if cfg.step_verify == "judge" {
            match llm_json(&spec.eval_spec.judge_model, STEP_JUDGE_SYSTEM,
                &format!("STEP:\n{}\n{}\ndone_when: {}\n\nOUTPUT:\n{}",
                    title, details, done_when,
                    output.chars().take(20000).collect::<String>()), 1000) {
                Ok(v) => (v["passed"].as_bool().unwrap_or(false),
                          v["note"].as_str().unwrap_or("").to_string(),
                          v["diagnosis"].as_str().unwrap_or("").to_string()),
                Err(e) => (false, String::new(), format!("judge failed: {}", e)),
            }
        } else {
            (true, output.chars().take(300).collect(), String::new())
        };

        if passed {
            state.steps[i].status = "done".into();
            state.steps[i].note = note.chars().take(500).collect();
            println!("  ✓ step {} checked off", i + 1);
        } else {
            println!("  ✗ step {} failed gate: {}", i + 1,
                     diagnosis.chars().take(120).collect::<String>());
            if attempts < cfg.max_attempts_per_step {
                state.steps[i].details.push_str(
                    &format!("\nPREVIOUS ATTEMPT FAILED — fix: {}",
                             diagnosis.chars().take(800).collect::<String>()));
            } else if cfg.replan_on_failure {
                println!("  [replan] revising remaining plan around the failure");
                let done_notes: String = state.steps.iter()
                    .filter(|s| s.status == "done")
                    .map(|s| format!("- {}: {}", s.title, s.note))
                    .collect::<Vec<_>>().join("\n");
                match llm_json(&cfg.planner_model, REPLAN_SYSTEM,
                    &format!("GOAL:\n{}\n\nCOMPLETED:\n{}\n\nFAILED STEP:\n{}\n\n\
                              DIAGNOSIS:\n{}", state.goal, done_notes, title,
                             diagnosis), 4000)
                    .and_then(|v| parse_steps(&v)) {
                    Ok(new_steps) => {
                        state.steps[i].status = "failed".into();
                        let mut kept: Vec<Step> = state.steps.iter()
                            .filter(|s| s.status != "pending").cloned().collect();
                        kept.extend(new_steps);
                        state.steps = kept;
                    }
                    Err(e) => {
                        println!("  [replan] failed ({}); marking step failed", e);
                        state.steps[i].status = "failed".into();
                    }
                }
            } else {
                state.steps[i].status = "failed".into();
            }
        }
        state.save(&state_path)?;          // resumability
        println!("\n{}\n", state.render());
    }
    println!("[loop] stopped: max_cycles ({}) reached", cfg.max_cycles);
    Ok(state)
}
